### PR TITLE
pkg/schema: create CamliType type

### PR DESCRIPTION
**Pull request by aviau**
pkg/schema: create CamliType type
    
    Create a CamliType type in pkg/schema and use it in a couple of
    packages.
    
    It can be implemented in other packages as we go.

I was reading pkg/index and decided to implement @mpl's TODO:
```
// TODO(mpl): create consts somewhere for "claim" and "permanode" as camliTypes, and use them,		if meta.CamliType != string(schema.PermanodeCamliType) && meta.CamliType != string(schema.ClaimCamliType) {
// instead of hardcoding. Unless they already exist ? (didn't find them).
```

### CODE REVIEW

File: pkg/schema/schema.go [hunk 1]
```diff
@@ -785,9 +785,27 @@ func populateParts(m map[string]interface{}, size int64, parts []BytesPart) erro
...
 func newBytes() *Builder {
-	return base(1, "bytes")
+	return base(1, BytesCamliType)
 }
 
+// CamliType is one of the valid "camliType" fields in a schema blob. See doc/schema
+type CamliType string
+
+const (
+	BytesCamliType     CamliType = "bytes"
```
Comment (bradfitz) [inline pkg/schema/schema.go:795]: How about "TypeBytes", "TypeClaim", etc, dropping "Camli" entirely from at least the const name?

File: pkg/search/handler.go [hunk 2]
```diff
@@ -808,14 +809,14 @@ func (h *Handler) EdgesTo(req *EdgesRequest) (*EdgesResponse, error) {
 		if found {
 			ei = &EdgeItem{
 				From:     edge.From,
-				FromType: "permanode",
+				FromType: string(schema.PermanodeCamliType),
```
Comment (bradfitz) [inline pkg/search/handler.go:812]: Why not change the EdgeItem.FromType type?

File: pkg/search/query.go [hunk 3]
```diff
@@ -383,7 +384,7 @@ func (c *Constraint) matchesAtMostOneBlob() blob.Ref {
 }
 
 func (c *Constraint) onlyMatchesPermanode() bool {
-	if c.Permanode != nil || c.CamliType == "permanode" {
+	if c.Permanode != nil || c.CamliType == string(schema.PermanodeCamliType) {
```
Comment (bradfitz) [inline pkg/search/query.go:387]: likewise, why not just the CamliType type?

File: pkg/index/corpus.go [hunk 4]
```diff
@@ -463,8 +463,8 @@ func (c *Corpus) scanFromStorage(s sorted.KeyValue) error {
 		return err
 	}
 
-	c.files = make(map[blob.Ref]camtypes.FileInfo, len(c.camBlobs["file"]))
-	c.permanodes = make(map[blob.Ref]*PermanodeMeta, len(c.camBlobs["permanode"]))
+	c.files = make(map[blob.Ref]camtypes.FileInfo, len(c.camBlobs[string(schema.FileCamliType)]))
```
Comment (bradfitz) [inline pkg/index/corpus.go:466]: likewise: camBlobs could be keyed by the new type?

**Review by bradfitz — COMMENTED** [review]
There are a lot of `string(foo)` in this change, which suggests something isn't fully right yet. See comments.

File: pkg/schema/schema.go (hunk 1 above)
Comment (aviau) [reply pkg/schema/schema.go]: Looks good, done.

File: pkg/search/handler.go (hunk 2 above)
Comment (aviau) [reply pkg/search/handler.go]: Done

File: pkg/search/query.go (hunk 3 above)
Comment (aviau) [reply pkg/search/query.go]: Done

File: pkg/index/corpus.go (hunk 4 above)
Comment (aviau) [reply pkg/index/corpus.go]: Done

**Comment by aviau** [comment]
I went the soft route at first, transforming it to strings all the time. However you are right we should just use CamliType everywhere.

I just pushed a new version of the patch that does just that. I also renamed it as you suggested.

**Comment by aviau** [comment]
@bradfitz Alright, done!

- `string` -> `CamliType` everywhere it made sense
- Renamed types to `Type<...>` as you suggested.

File: pkg/index/corpus.go [hunk 5]
```diff
@@ -675,7 +675,7 @@ func (c *Corpus) mergeBlobMeta(bm camtypes.BlobMeta) error {
 	if _, dup := c.blobs[bm.Ref]; dup {
 		panic("dup blob seen")
 	}
-	bm.CamliType = c.str(bm.CamliType)
+	bm.CamliType = schema.CamliType((c.str(string(bm.CamliType))))
```
Comment (aviau) [inline pkg/index/corpus.go:678]: Not really sure if that works, can you confirm?

Comment (bradfitz) [reply pkg/index/corpus.go:678]: converting between two string-backed types is free. they'll share the same memory.

File: pkg/server/download.go [hunk 6]
```diff
@@ -416,7 +416,7 @@ func (dh *DownloadHandler) statFiles(refs []blob.Ref) error {
 	return nil
 }
 
-var allowedFileTypes = map[string]bool{"file": true, "symlink": true, "fifo": true, "socket": true}
+var allowedFileTypes = map[schema.CamliType]bool{schema.TypeFile: true, schema.TypeSymlink: true, schema.TypeFIFO: true, schema.TypeSocket: true}
```
Comment (bradfitz) [inline pkg/server/download.go:419]: I'd break this up onto 6 lines

File: pkg/schema/schema.go [hunk 7]
```diff
@@ -785,9 +785,27 @@ func populateParts(m map[string]interface{}, size int64, parts []BytesPart) erro
 }
 
 func newBytes() *Builder {
-	return base(1, "bytes")
+	return base(1, TypeBytes)
 }
 
+// CamliType is one of the valid "camliType" fields in a schema blob. See doc/schema
```
Comment (bradfitz) [inline pkg/schema/schema.go:791]: end in period.

**Review by bradfitz — APPROVED** [review]

File: pkg/server/download.go (hunk 6 above)
Comment (aviau) [reply pkg/server/download.go]: Yeah, much nicer.

File: pkg/schema/schema.go (hunk 7 above)
Comment (aviau) [reply pkg/schema/schema.go]: done!

**Pull request merged**